pkg/controller/node/nodecontroller.go

```diff
@@ -926,3 +1083,139 @@ func (nc *NodeController) ComputeZoneState(nodeReadyConditions []*api.NodeCondit
 		return notReadyNodes, stateNormal
 	}
 }
+
+type modifyTaintsFunc func(oldTaints []api.Taint) ([]api.Taint, bool, error)
+
+func tryModifyNodeTaints(kubeClient clientset.Interface, nodeName string, modifyTaints modifyTaintsFunc) (bool, error) {
+	node, err := kubeClient.Core().Nodes().Get(nodeName)
+	if err != nil {
+		return false, err
+	}
+
+	oldTaints, err := api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		return false, err
+	}
+
+	newTaints, updated, err := modifyTaints(oldTaints)
+	if err != nil || !updated {
+		return false, err
+	}
+
+	taintsData, err := json.Marshal(newTaints)
+	if err != nil {
+		return false, err
+	}
+
+	if len(node.Annotations) == 0 {
+		node.Annotations = map[string]string{}
+	}
+	node.Annotations[api.TaintsAnnotationKey] = string(taintsData)
+
+	_, err = kubeClient.Core().Nodes().Update(node)
+	return err == nil, err
+}
+
+func tryAddTaintToNode(kubeClient clientset.Interface, nodeName string, taintToAdd *api.Taint) (bool, error) {
+	addTaints := func(oldTaints []api.Taint) ([]api.Taint, bool, error) {
+		newTaints := []api.Taint{}
+		for _, oldTaint := range oldTaints {
+			if taintToAdd.MatchTaint(oldTaint) {
+				return []api.Taint{}, false, nil
+			}
+		}
+		newTaints = append(oldTaints, *taintToAdd)
+		return newTaints, true, nil
+	}
+
+	return tryModifyNodeTaints(kubeClient, nodeName, addTaints)
+}
+
+// tryRemoveTaintsOffNode tries to remove taints off a node,
+// return true if taints removed, or return false if taints don't exist.
+func tryRemoveTaintsOffNode(kubeClient clientset.Interface, nodeName string, taintsToRemove ...api.Taint) (bool, error) {
+	removeTaints := func(taints []api.Taint) ([]api.Taint, bool, error) {
+		removed := false
+		for _, taintToRemove := range taintsToRemove {
+			updated := false
+			taints, updated = api.DeleteTaint(taints, taintToRemove)
+			removed = removed || updated
+		}
+
+		return taints, removed, nil
+	}
+
+	return tryModifyNodeTaints(kubeClient, nodeName, removeTaints)
+}
+
+// trySyncNodeTaintsWithNodeStatus will try to update node taint according to node status.
+// If node is observed not ready (NodeReady status ConditionFalse), add taint `notready:NoExecute` to node,
+// if node is observed unknown (NodeReady status ConditionUnknown), add taint `unreachable:NoExecute` to node,
+// if node is observed ready, remove these two taints.
+// TODO: may incorporate these into tryUpdateNodeStatus to reduce api call when taint becomes a field of nodeStatus
+func (nc *NodeController) trySyncNodeTaintsWithNodeStatus(nodeName string, decisionTimestamp unversioned.Time, observedReadyCondition, currentReadyCondition *api.NodeCondition) {
```

Comment:
This whole logic is pretty much copy of what's being done in `tryUpdateNodeStatus`. We shouldn't do that. Instead please incorporate this into that function. IIUC the logic we want to have is that in all places where we updated NodeStatus we should now update Taints as well, and when deciding on eviction we should, depending on flag, either test `transitionTimestamp` against `podEvictionTimeout` (old behavior) or `taintAge` (however we choose to name it) against shortest toleration across all Pods assigned to the given Node (it's bad, but I don't have a better idea - we just need to process all Pods assigned to a Node every time it gets not Ready for any reason).

It seems pretty expensive though... @wojtek-t